### ISSUE TITLE
Error trying tensorflow litem operations are not supported by GPU delegate

### ISSUE THREAD

**Issue by lwu025** [tensorflow/tensorflow#29086]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 16.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: Yes, aarch64, android 8.1
- TensorFlow installed from (source or binary): source
- TensorFlow version (use command below): 1.12.2
- Python version: 2.7
- Bazel version (if compiling from source): 0.22.0
- GCC/Compiler version (if compiling from source): 
- CUDA/cuDNN version:
- GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
I'm trying to run a c++ demo of tflite+opengles on my aarch64 android 8.1 board.
I have tried to build  benchmark_model tool with bazel, it ran the [deeplabv3_257_mv_gpu.tflite ](https://storage.googleapis.com/download.tensorflow.org/models/tflite/gpu/deeplabv3_257_mv_gpu.tflite)model successfully on my device.

Now I want to integrate a simple demo of tflite (similar to benchmark) to my codes, which is built by cmake. I spent some time extract all the static libs of tensorflowlite from bazle-out folder and linked them in my cmake. I built the code successfully with ndk standalone toolchain r17c. 

But when I ran this new demo on my device it shows me errors like:

`INFO: Created TensorFlow Lite delegate for GPU.
Apply delegate for GPU
Next operations are not supported by GPU delegate:
AVERAGE_POOL_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
CONV_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
RESIZE_BILINEAR: Expected 1 input tensor(s), but node has 2 runtime input(s).
First 1 operations will run on the GPU, and the remaining 69 on the CPU.
TfLiteGpuDelegate Prepare: ReadValue: value is a constant tensor: 183
Node number 70 (TfLiteGpuDelegate) failed to prepare.

Failed to apply GPU delegate.
Delegate setting done
Node number 70 (TfLiteGpuDelegate) failed to prepare.

Failed to allocate tensors!
`
The model here is still deeplabv3_257_mv_gpu.tflite, which have been proved working on my device. I also have tried to build my new demo code in bazel, which perform correctly.

**Describe the expected behavior**
Build Tflite + opengl delegate successfully with Cmake and perform correctly on my aarch64 board. 

**Code to reproduce the issue**
Provide a reproducible test case that is the bare minimum necessary to generate the problem.
use your official benchmark_model.cc

**Other info / logs**
Include any logs or source code that would be helpful to diagnose the problem. If including tracebacks, please include the full traceback. Large logs and files should be attached.

**Comment by impjdi** [comment]
@lwu025 

The shader code is very sensitive to what a runtime tensor is and what a constant weight / bias is.  You might have changed some constant tensors as runtime tensors so that you can easily inspect tensor content, but that's not compatible with the GPU backend at this point.  From:

> AVERAGE_POOL_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
> CONV_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
> CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
> CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
> DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
> DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
> RESIZE_BILINEAR: Expected 1 input tensor(s), but node has 2 runtime input(s).

shows that there's bunch of wrongly connected things, e.g. what's an `AVERAGE_POOL_2D` or `CONV_2D` useful if there's no runtime tensor?

**Comment by lwu025** [comment]
Thanks. I rebuild the tensorflow lite with the following modifications:
Add the following code to tensorflow/lite/BUILD
`cc_binary(
	name = "libtensorflow-lite.so",
	linkopts=[
		"-shared",
		"-Wl,-soname=libtensorflow-lite.so",
		],
		linkshared = 1,
		copts = tflite_copts(),
		deps = [
		":framework",
		"//tensorflow/lite/kernels:builtin_ops",
	] + select({
        "//tensorflow:android_arm64": ["//tensorflow/lite/delegates/gpu:gl_delegate"],
        "//conditions:default": [],
    }),
)`

and build command:
`sudo bazel build //tensorflow/lite:libtensorflow-lite.so --crosstool_top=//external:android/crosstool --cpu=arm64-v8a --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --cxxopt="-std=c++14" --experimental_better_python_version_mixing
`

I followed the sample code in delegate/gpu/readme.md and my demo works good!

I have one more question here:
Is it possible to build the tensorflow lite lib for aarch64 linux (NO android!)?
If yes, how to build it in bazel?

**Comment by impjdi** [comment]
@lwu025 

Hm, good question.  I haven't built anything else but x86, android_arm64, and ios_arm64.  So, I don't know the answer to that question.

re: GPU running on those platforms, we haven't tried, and thus we don't know either.  One comment I can add though is that we were running the GPU delegate on x86 desktop with mesa for unit tests.  So maybe something equivalent exists for aarch64, but I'm not sure =/

Given threads like https://github.com/tensorflow/tensorflow/issues/22629 I think it should be possible.  You should be able to build things with https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/BUILD and https://github.com/tensorflow/tensorflow/blob/master/tensorflow/lite/delegates/gpu/BUILD should mesa-equivalent thing run for your aarch64.

**Comment by weinixuehao** [comment]
@lwu025 tflite gpu only ios arm64 and x86_64?not support armv7 armv7s?
I need armv7 armv7a as my ios project 
![image](https://user-images.githubusercontent.com/17869361/60001400-f9b38d00-9698-11e9-9f7d-9c1072e582be.png)

**Comment by lwu025** [comment]
@weinixuehao I never try armv7 before. You can try it simply replay the --cpu option when you build tflite.

**Comment by sirius0503** [comment]
> @lwu025
> 
> The shader code is very sensitive to what a runtime tensor is and what a constant weight / bias is. You might have changed some constant tensors as runtime tensors so that you can easily inspect tensor content, but that's not compatible with the GPU backend at this point. From:
> 
> > AVERAGE_POOL_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
> > CONV_2D: Expected 1 input tensor(s), but node has 0 runtime input(s).
> > CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
> > CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
> > DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
> > DEPTHWISE_CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
> > RESIZE_BILINEAR: Expected 1 input tensor(s), but node has 2 runtime input(s).
> 
> shows that there's bunch of wrongly connected things, e.g. what's an `AVERAGE_POOL_2D` or `CONV_2D` useful if there's no runtime tensor?

@impjdi : Can you explain this better by dumbing it down a bit, I have the same error, the **tflite model** works great on my pc with the python interpreter, but on android gives the following error:

```
java.lang.IllegalStateException: Internal error: Unexpected failure when preparing tensor allocations: Next operations are not supported by GPU delegate:
2020-02-25 15:15:57.160 22863-22863/com.valuepitch.intruderdetector W/System.err: CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).
2020-02-25 15:15:57.161 22863-22863/com.valuepitch.intruderdetector W/System.err: CONV_2D: Expected 1 input tensor(s), but node has 3 runtime input(s).
2020-02-25 15:15:57.161 22863-22863/com.valuepitch.intruderdetector W/System.err: L2_NORMALIZATION: Operation is not supported.
2020-02-25 15:15:57.161 22863-22863/com.valuepitch.intruderdetector W/System.err: MEAN: Expected 1 input tensor(s), but node has 2 runtime input(s).
2020-02-25 15:15:57.161 22863-22863/com.valuepitch.intruderdetector W/System.err: SUM: Operation is not supported.
2020-02-25 15:15:57.162 22863-22863/com.valuepitch.intruderdetector W/System.err: TOPK_V2: Operation is not supported.
2020-02-25 15:15:57.162 22863-22863/com.valuepitch.intruderdetector W/System.err: First 0 operations will run on the GPU, and the remaining 186 on th
2020-02-25 15:15:57.163 22863-22863/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.NativeInterpreterWrapper.allocateTensors(Native Method)
2020-02-25 15:15:57.163 22863-22863/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.NativeInterpreterWrapper.run(NativeInterpreterWrapper.java:145)
2020-02-25 15:15:57.164 22863-22863/com.valuepitch.intruderdetector W/System.err:     at org.tensorflow.lite.Interpreter.runForMultipleInputsOutputs(Interpreter.java:311)
2020-02-25 15:15:57.164 22863-22863/com.valuepitch.intruderdetector W/System.err:     at com.valuepitch.intruderdetector.RunModel.justnewTest(RunModel.java:723)
2020-02-25 15:15:57.164 22863-22863/com.valuepitch.intruderdetector W/System.err:     at com.valuepitch.intruderdetector.RunModel.access$100(RunModel.java:43)
```

**Comment by sirius0503** [comment]
> Thanks. I rebuild the tensorflow lite with the following modifications:
> Add the following code to tensorflow/lite/BUILD
> `cc_binary( name = "libtensorflow-lite.so", linkopts=[ "-shared", "-Wl,-soname=libtensorflow-lite.so", ], linkshared = 1, copts = tflite_copts(), deps = [ ":framework", "//tensorflow/lite/kernels:builtin_ops", ] + select({ "//tensorflow:android_arm64": ["//tensorflow/lite/delegates/gpu:gl_delegate"], "//conditions:default": [], }), )`
> 
> and build command:
> `sudo bazel build //tensorflow/lite:libtensorflow-lite.so --crosstool_top=//external:android/crosstool --cpu=arm64-v8a --host_crosstool_top=@bazel_tools//tools/cpp:toolchain --cxxopt="-std=c++14" --experimental_better_python_version_mixing `
> 
> I followed the sample code in delegate/gpu/readme.md and my demo works good!
> 
> I have one more question here:
> Is it possible to build the tensorflow lite lib for aarch64 linux (NO android!)?
> If yes, how to build it in bazel?

@lwu025 : Does this remove the ` CONV_2D: Expected 1 input tensor(s), but node has 2 runtime input(s).` type errors?

**Comment by impjdi** [comment]
@sirius0503 

Yeah, as you noted, the properties of the input tensors, the number of the input tensors, and the order of the input tensors is very important for the GPU delegate.  So for example, if you are trying to feed in some constant tensor just to see whether things produce the expected number, you're not gonna have a good time, because you'll get something like:

> Expected 1 input tensor(s), but node has 0 runtime input(s).

You should properly have a placeholder and fill it in, rather than feeding a constant tensor.

Also, it's the same with the weights and biases.  They are supposed to be holding the content after training, so they should be read only constant tensors.  If you want to check whether things work, and make them runtime tensors, you will start seeing messages like:

> Expected 1 input tensor(s), but node has 2 runtime input(s).

`MEAN` doesn't compute the mean of multiple tensors, but just inside a single tensor.  It's added for instance normalization.  We probably haven't studied `MEAN` too well and didn't anticipate that use case (maybe it's a valid thing in TFLite).

I don't think we ever implemented `L2_NORMALIZATION`, `SUM`, and `TOPK_V2`.

**Comment by sirius0503** [comment]
@impjdi : I have used keras `input layer` for feeding inputs, and not a `tf.constant`, also the `CONV_2D` comes from my `pretrained model`, so I think as you mentioned, I have made my weights and biases as runtime tensors ( also what are runtime tensors) and how did my weights and biases become runtime tensors, can you explain. ( I don't have a clue). Also , `tf.nn.top_k, tf.nn.l2_normalization ` are supported for **tflite**, when you say there isn't an implementation are you saying a **gpu implementation**. Also, thanks for the great explanation! Really helped! :)

> Yeah, as you noted, the properties of the input tensors, the number of the input tensors, and the order of the input tensors is very important for the GPU delegate.

Can you also explain the above part : specifically how does _number of inputs, and order of the input tensors_ impact the GPU delegate?

**Comment by impjdi** [comment]
@sirius0503 

> how did my weights and biases become runtime tensors, can you explain.

Uh, no idea.  I am not really familiar with training & converting; I usually only work with finished TFLite files.  I wonder whether you set up the TF graph wrong, so that it propagated all the way down to TFLite.  Or it could be that there is a bug in TOCO (which I doubt, as everyone must be screaming by now).

> are you saying a gpu implementation

That is correct.  GPU shader implementation :)

> number of inputs

As a hypothetical example, our `ADD` implementation currently only takes in 1 or 2 input tensors.  I think in TFLite, you can do more, but with TFLite GPU, if you do more, it will probably say it can't handle it.

> order of the input tensors

As an example, the input tensors of `CONV_2D` must be 0: input tensor, 1: weight tensor, and 2: optional bias tensor.  You can't change thh order to, e.g., 0: bias tensor, 1: input tensor, 2: weight tensor, and hope that GPU can figure out the stuff by its own by looking at the tensor dimensions.

**Comment by sirius0503** [comment]
@impjdi : 

>  how did my weights and biases become runtime tensors, can you explain.

Can somebody else explain why this is happening ? I can't say whether I set the TF graph wrong. If someone can give me an example of such a wrong graph, I can better understand where I'm making mistakes.

Also, I used the command line `tflite_convert` and not **TOCO** or the `python api` can it make a difference?

**Comment by impjdi** [comment]
@sirius0503 

You froze the graph, right?

**Comment by yxchng** [comment]
@impjdi I am facing this error for simple addition and subtraction but I am unable to understand your comment of 

> The shader code is very sensitive to what a runtime tensor is and what a constant weight / bias is. You might have changed some constant tensors as runtime tensors so that you can easily inspect tensor content, but that's not compatible with the GPU backend at this point. 

Can you explain in more details what it means and how to avoid such errors?

**Comment by impjdi** [comment]
A runtime tensor is a tensor that is computed with at least another.  The very first runtime tensor is possibly the input tensor of the model.

A constant tensor is a tensor that is fixed.  These are, but not limited to, trained weights and biases for conv/fully connected ops or constants for add/mul or parameters for reshape etc.

If you think about the formula y = a * x + b, x and y are runtime tensors (variables) and a and b are constant tensors (constants).

**Comment by tensorflowbutler** [comment]
Hi There,

 We are checking to see if you still need help on this issue, as you are using an older version of tensorflow(1.x) which is officially considered as end of life. We recommend that you upgrade to 2.4 or later version and let us know if the issue still persists in newer versions. 

 This issue will be closed automatically 7 days from now. If you still need help with this issue, Please open a new issue for any help you need against 2.x, and we will get you the right help.
